File: yht_custom/api/payment_assist.py
# Copyright (c) 2026, Enfono Technologies and contributors
# For license information, please see license.txt

"""Sales-side payment assist — MoM 5.6.

Three jobs, all read-only except the last:

1. ``get_branch_payment_modes`` — which Modes of Payment this branch may tender
   against, resolved server-side.
2. ``get_customer_payment_status`` — outstanding, credit limit and the breach
   flag, so the operator sees the customer's position before committing a credit
   sale.
3. ``collect_payment`` — turn a tender split into submitted Payment Entries.

WHY THIS DOES NOT CALL ``get_payment_entry``
--------------------------------------------
RMAX's equivalent does, and on a branch user it dies with a bare
``PermissionError`` and no message. Traced through v15 on this bench:

    payment_entry.py:2961   pe.bank_account = <default company Bank Account>
    payment_entry.py:2973   pe.set_bank_account_data()
    payment_entry.py:254      if not self.bank_account: return     # the escape
    bank_account.py:142       frappe.has_permission("Bank Account", ptype="read",
                                                    doc=..., throw=True)

The check is only reached because ``get_payment_entry`` populates
``pe.bank_account`` from the company's default. On this site
``Saudi National Bank - KATC`` carries ``is_default = 1``, so it IS populated and
the throw DOES fire. ``get_bank_account_details`` re-reads with a hard
``throw=True``, which no ``ignore_permissions`` flag suppresses.

Granting ``read`` on Bank Account would clear it in one line, but Bank Account has
no permlevel split — ``read`` hands every branch operator the account number and
IBAN of every company account. So the Payment Entry is built here field by field
instead. Nothing in this module reads Bank Account, and nothing it returns to the
browser contains an account number, an IBAN or a GL account code: the dialog is
labelled with Mode of Payment names only.

CREDIT NOTES
------------
Returns are excluded deliberately. RMAX's tender dialog fires on every draft with
``custom_payment_mode == "Cash"`` and never checks ``is_return``, so it also
appears on credit notes and then refuses to let them save until the full amount is
allocated — logged as an open defect in RMAX's own ROADMAP.md. This site already
holds 99 returns, so the bug would land immediately. ``collect_payment`` throws on
a return and the client gate skips it.
"""

import json

import frappe
from frappe import _
from frappe.query_builder.functions import Sum
from frappe.utils import flt, getdate

#: Roles that see every enabled mode, not just the branch's allowlist.
_UNRESTRICTED_ROLES = {
	"System Manager",
	"Accounts Manager",
	"Accounts User",
	"Sales Manager",
	"Sales Master Manager",
}


def _is_unrestricted(user: str | None = None) -> bool:
	user = user or frappe.session.user
	if user == "Administrator":
		return True
	roles = set(frappe.get_roles(user))
	if roles & _UNRESTRICTED_ROLES:
		return True
	return "Branch User" not in roles


def _user_branch(user: str | None = None) -> str | None:
	"""The Branch Configuration row this user belongs to, if any."""
	user = user or frappe.session.user
	parents = frappe.get_all(
		"Branch Configuration User", filters={"user": user}, pluck="parent", limit=1
	)
	return parents[0] if parents else None


def _modes_with_account(company: str) -> set[str]:
	"""Modes carrying a default Cash/Bank account for this company.

	A mode without one cannot post: ``get_bank_cash_account`` throws. Offering it
	in the dialog would mean the operator allocates money and only finds out at
	submit. Measured on this site: 21 modes enabled, 8 usable.
	"""
	mode_account = frappe.qb.DocType("Mode of Payment Account")
	rows = (
		frappe.qb.from_(mode_account)
		.select(mode_account.parent)
		.distinct()
		.where(
			(mode_account.company == company)
			& mode_account.default_account.notnull()
			& (mode_account.default_account != "")
		)
	).run()
	return {r[0] for r in rows}


@frappe.whitelist()
def get_branch_payment_modes(company: str) -> list[dict]:
	"""Tenderable modes for the signed-in user.

	Returns ``[{"mode_of_payment": str, "type": str}]`` — names and Cash/Bank type
	only. No account, no account number, no IBAN.
	"""
	if not company:
		return []

	usable = _modes_with_account(company)
	if not usable:
		return []

	allowed = None
	if not _is_unrestricted():
		branch = _user_branch()
		if branch:
			listed = frappe.get_all(
				"Branch Configuration Mode of Payment",
				filters={"parent": branch},
				pluck="mode_of_payment",
			)
			# An unconfigured branch opts out of filtering rather than losing every
			# mode — the same convention Branch Configuration uses elsewhere. The
			# moment one row is listed, the allowlist is authoritative.
			if listed:
				allowed = set(listed)

	modes = frappe.get_all(
		"Mode of Payment",
		filters={"enabled": 1, "name": ["in", list(usable)]},
		fields=["name", "type"],
		order_by="type desc, name asc",
	)

	out = []
	for mode in modes:
		if allowed is not None and mode.name not in allowed:
			continue
		out.append({"mode_of_payment": mode.name, "type": mode.type or ""})
	return out


@frappe.whitelist()
def get_customer_payment_status(customer: str, company: str) -> dict:
	"""Outstanding, credit limit and breach flag for the sales screen.

	``frappe.has_permission`` is checked explicitly: this is a public HTTP endpoint
	and the answer leaks a customer's total exposure.
	"""
	if not customer or not company:
		return {}

	if not frappe.has_permission("Customer", "read", doc=customer):
		raise frappe.PermissionError

	invoice = frappe.qb.DocType("Sales Invoice")
	rows = (
		frappe.qb.from_(invoice)
		.select(Sum(invoice.outstanding_amount))
		.where(
			(invoice.docstatus == 1)
			& (invoice.customer == customer)
			& (invoice.company == company)
		)
	).run()
	outstanding = flt(rows[0][0]) if rows and rows[0] else 0.0

	limit_row = frappe.db.get_value(
		"Customer Credit Limit",
		{"parent": customer, "company": company},
		["credit_limit", "bypass_credit_limit_check"],
		as_dict=True,
	)
	credit_limit = flt(limit_row.credit_limit) if limit_row else 0.0
	bypass = bool(limit_row and limit_row.bypass_credit_limit_check)

	return {
		"customer": customer,
		"outstanding": outstanding,
		"credit_limit": credit_limit,
		"bypass_credit_limit_check": bypass,
		# No limit set is NOT a breach. Measured on this site: 0 Customer Credit
		# Limit rows exist, so this stays False everywhere until YHT sets limits
		# (blocker B5). The warning is wired and dormant, not missing.
		"has_limit": credit_limit > 0,
	}


def _resolve_mode_account(mode_of_payment: str, company: str) -> str:
	"""The GL account a mode posts to. Server-side only — never returned."""
	account = frappe.db.get_value(
		"Mode of Payment Account",
		{"parent": mode_of_payment, "company": company},
		"default_account",
	)
	if not account:
		frappe.throw(
			_("Mode of Payment {0} has no default account for {1}.").format(
				frappe.bold(mode_of_payment), frappe.bold(company)
			)
		)
	return account


@frappe.whitelist()
def collect_payment(sales_invoice: str, payments: str | list) -> list[str]:
	"""Create and submit one Payment Entry per tendered mode.

	``payments``: ``[{"mode_of_payment": str, "amount": float}, ...]``.
	Rows with a non-positive amount are dropped, not rejected — the dialog shows
	every branch mode and most stay at 0.00.
	"""
	if not sales_invoice:
		frappe.throw(_("Sales Invoice is required."))

	invoice = frappe.get_doc("Sales Invoice", sales_invoice)

	# get_doc does not check read permission on its own.
	if not frappe.has_permission("Sales Invoice", "read", doc=invoice):
		raise frappe.PermissionError
	if not frappe.has_permission("Payment Entry", "create"):
		raise frappe.PermissionError

	if invoice.docstatus != 1:
		frappe.throw(
			_("{0} must be submitted before payment can be recorded.").format(invoice.name)
		)
	if invoice.is_return:
		# See the module docstring — a credit note is not a collection.
		frappe.throw(
			_("{0} is a credit note. Record a refund as a Payment Entry instead.").format(
				invoice.name
			)
		)

	if isinstance(payments, str):
		try:
			payments = json.loads(payments)
		except (ValueError, TypeError):
			frappe.throw(_("Could not read the payment rows."))
	if not isinstance(payments, (list, tuple)):
		frappe.throw(_("Could not read the payment rows."))

	rows = []
	for row in payments:
		mode = (row or {}).get("mode_of_payment")
		amount = flt((row or {}).get("amount"))
		if not mode or amount <= 0:
			continue
		rows.append({"mode_of_payment": mode, "amount": amount})

	if not rows:
		frappe.throw(_("Enter an amount against at least one mode of payment."))

	allowed = {m["mode_of_payment"] for m in get_branch_payment_modes(invoice.company)}
	for row in rows:
		if row["mode_of_payment"] not in allowed:
			# The dialog only ever offers the allowlist; anything else arrived by
			# hand-crafting the request.
			frappe.throw(
				_("Mode of Payment {0} is not available for your branch.").format(
					frappe.bold(row["mode_of_payment"])
				)
			)

	tendered = flt(sum(r["amount"] for r in rows), 2)
	outstanding = flt(invoice.outstanding_amount, 2)
	if tendered - outstanding > 0.01:
		frappe.throw(
			_("Tendered {0} is more than the outstanding {1} on {2}.").format(
				frappe.bold(tendered), frappe.bold(outstanding), invoice.name
			)
		)

	created = []
	for row in rows:
		created.append(_make_payment_entry(invoice, row["mode_of_payment"], row["amount"]))
	return created


def _make_payment_entry(invoice, mode_of_payment: str, amount: float) -> str:
	"""Build a Receive Payment Entry against one invoice, one mode.

	Written out longhand rather than via ``get_payment_entry`` so no Bank Account
	read is required — see the module docstring. Only the fields a single-currency
	customer receipt needs are set; the controller derives the rest.
	"""
	deposit_account = _resolve_mode_account(mode_of_payment, invoice.company)
	receivable = invoice.debit_to
	party_currency = frappe.get_cached_value("Account", receivable, "account_currency")
	deposit_currency = frappe.get_cached_value("Account", deposit_account, "account_currency")

	if party_currency != deposit_currency:
		# Out of scope: YHT trades in SAR only and every price list on this site is
		# SAR. Fail loudly rather than post a wrong exchange rate.
		frappe.throw(
			_("{0} and {1} are in different currencies. Record this payment manually.").format(
				frappe.bold(mode_of_payment), frappe.bold(invoice.currency)
			)
		)

	entry = frappe.new_doc("Payment Entry")
	entry.payment_type = "Receive"
	entry.company = invoice.company
	entry.posting_date = getdate(invoice.posting_date)
	entry.mode_of_payment = mode_of_payment
	entry.party_type = "Customer"
	entry.party = invoice.customer
	entry.paid_from = receivable
	entry.paid_from_account_currency = party_currency
	entry.paid_to = deposit_account
	entry.paid_to_account_currency = deposit_currency
	entry.paid_amount = amount
	entry.received_amount = amount
	entry.source_exchange_rate = 1
	entry.target_exchange_rate = 1
	entry.reference_no = invoice.name
	entry.reference_date = getdate(invoice.posting_date)
	if invoice.get("cost_center"):
		entry.cost_center = invoice.cost_center

	entry.append(
		"references",
		{
			"reference_doctype": "Sales Invoice",
			"reference_name": invoice.name,
			"due_date": invoice.get("due_date"),
			"total_amount": flt(invoice.grand_total),
			"outstanding_amount": flt(invoice.outstanding_amount),
			"allocated_amount": amount,
		},
	)

	entry.insert()
	entry.submit()
	return entry.name

File: yht_custom/public/js/payment_assist.js
// Copyright (c) 2026, Enfono Technologies and contributors
// For license information, please see license.txt
//
// Sales-side payment assist (MoM §5.6): the cash/credit decision, the customer's
// outstanding position, the credit-limit warning, and the tender dialog.
//
// TWO DELIBERATE DIFFERENCES FROM RMAX, both to avoid known defects:
//
//   1. THE DIALOG FIRES AFTER SUBMIT, NOT AT SAVE.
//      RMAX opens it on a draft and then refuses to let the document save until
//      the full amount is allocated. That is wrong whenever the operator does not
//      yet know the split, and there is no discoverable escape except changing
//      Payment Mode away from Cash. Here the invoice saves and submits normally;
//      the dialog opens once there is something real to pay against, and closing
//      it leaves a perfectly valid unpaid invoice. `Collect Payment` in the
//      toolbar reopens it any time.
//
//   2. RETURNS ARE EXCLUDED.
//      RMAX's gate never checks `is_return` — the string does not appear in its
//      popup file — so the dialog also appears on credit notes, labelled
//      "Enter Payment Amounts" against a positive "Invoice Total" while the money
//      is actually going out. Logged as an open defect in RMAX's own ROADMAP.md.
//      This site already holds 99 returns, so it would land on day one.
//
// The server never sends an account number, an IBAN or a GL account code to this
// file — see yht_custom/api/payment_assist.py for why that mattered.

frappe.provide("yht_custom.payment");

frappe.ui.form.on("Sales Invoice", {
	refresh(frm) {
		yht_custom.payment.show_customer_position(frm);
		yht_custom.payment.add_collect_button(frm);
	},

	customer(frm) {
		yht_custom.payment.show_customer_position(frm);
	},

	custom_payment_mode(frm) {
		yht_custom.payment.show_customer_position(frm);
	},

	after_submit(frm) {
		if (!yht_custom.payment.wants_tender(frm)) return;
		// A beat, so the submit alert and the docstatus re-render land first.
		setTimeout(() => yht_custom.payment.open_tender(frm), 400);
	},

	before_save(frm) {
		return yht_custom.payment.warn_over_credit_limit(frm);
	},
});

// ------------------------------------------------------------ customer position

yht_custom.payment.status_cache = {};

yht_custom.payment.fetch_status = function (frm) {
	if (!frm.doc.customer || !frm.doc.company) return Promise.resolve(null);
	const key = `${frm.doc.customer}::${frm.doc.company}`;
	if (yht_custom.payment.status_cache[key]) {
		return Promise.resolve(yht_custom.payment.status_cache[key]);
	}
	return frappe
		.call({
			method: "yht_custom.api.payment_assist.get_customer_payment_status",
			args: { customer: frm.doc.customer, company: frm.doc.company },
		})
		.then((r) => {
			const status = r.message || null;
			if (status) yht_custom.payment.status_cache[key] = status;
			return status;
		});
};

// The customer's exposure, on the sales screen, before the sale is committed.
yht_custom.payment.show_customer_position = function (frm) {
	if (!frm.doc.customer) return;
	yht_custom.payment.fetch_status(frm).then((status) => {
		if (!status) return;
		const money = (v) => format_currency(v || 0, frm.doc.currency);

		const outstanding = flt(status.outstanding);
		frm.dashboard.add_indicator(
			__("Outstanding: {0}", [money(outstanding)]),
			outstanding > 0 ? "orange" : "green"
		);

		if (status.has_limit) {
			const exposure = outstanding + flt(frm.doc.grand_total);
			const over = exposure - flt(status.credit_limit);
			frm.dashboard.add_indicator(
				__("Credit limit: {0}", [money(status.credit_limit)]),
				over > 0 ? "red" : "blue"
			);
		}
	});
};

// ------------------------------------------------------------- credit-limit gate
//
// B5 was answered "warn, allow override", so this resolves rather than rejects.
// `bypass_credit_limit_check` on the customer's own limit row silences it, which
// is ERPNext's existing convention for the same decision.
//
// Dormant until limits exist: measured 0 Customer Credit Limit rows on this site.

yht_custom.payment.warn_over_credit_limit = function (frm) {
	if (frm.doc.custom_payment_mode !== "Credit") return;
	if (frm.doc.is_return) return;
	if (!frm.doc.customer || !flt(frm.doc.grand_total)) return;

	return yht_custom.payment.fetch_status(frm).then((status) => {
		if (!status || !status.has_limit || status.bypass_credit_limit_check) return;

		const money = (v) => format_currency(v || 0, frm.doc.currency);
		// The invoice's own total counts only while it is still a draft; once
		// submitted it is already inside `outstanding`.
		const pending = frm.doc.docstatus === 0 ? flt(frm.doc.grand_total) : 0;
		const exposure = flt(status.outstanding) + pending;
		if (exposure <= flt(status.credit_limit)) return;

		frappe.show_alert(
			{
				message: __("{0} would be {1} over their credit limit.", [
					frappe.utils.escape_html(frm.doc.customer_name || frm.doc.customer),
					money(exposure - flt(status.credit_limit)),
				]),
				indicator: "red",
			},
			10
		);
	});
};

// ------------------------------------------------------------------- tender flow

yht_custom.payment.wants_tender = function (frm) {
	return (
		frm.doc.custom_payment_mode === "Cash" &&
		!frm.doc.is_return &&
		frm.doc.docstatus === 1 &&
		flt(frm.doc.outstanding_amount) > 0
	);
};

yht_custom.payment.add_collect_button = function (frm) {
	if (!yht_custom.payment.wants_tender(frm)) return;
	frm.add_custom_button(__("Collect Payment"), () => yht_custom.payment.open_tender(frm));
};

yht_custom.payment.open_tender = function (frm) {
	frappe.call({
		method: "yht_custom.api.payment_assist.get_branch_payment_modes",
		args: { company: frm.doc.company },
		callback(r) {
			const modes = r.message || [];
			if (!modes.length) {
				frappe.msgprint({
					title: __("No payment modes available"),
					message: __(
						"No Mode of Payment is set up with a default account for {0}, or none is allowed for your branch. Ask your administrator to configure the branch's payment modes.",
						[frappe.utils.escape_html(frm.doc.company)]
					),
					indicator: "orange",
				});
				return;
			}
			yht_custom.payment.tender_dialog(frm, modes);
		},
	});
};

yht_custom.payment.tender_dialog = function (frm, modes) {
	const outstanding = flt(frm.doc.outstanding_amount);
	const money = (v) => format_currency(v || 0, frm.doc.currency);

	const fields = [
		{
			fieldtype: "Currency",
			fieldname: "outstanding",
			label: __("Amount Due"),
			default: outstanding,
			read_only: 1,
		},
		{ fieldtype: "Section Break" },
	];

	// One row per mode. Cash types first — that is what a counter reaches for.
	modes.forEach((mode, i) => {
		fields.push({
			fieldtype: "Currency",
			fieldname: `amount_${i}`,
			label: mode.mode_of_payment,
			default: 0,
		});
	});

	fields.push(
		{ fieldtype: "Section Break" },
		{ fieldtype: "HTML", fieldname: "summary" }
	);

	const dialog = new frappe.ui.Dialog({
		title: __("Collect Payment"),
		fields: fields,
		primary_action_label: __("Record Payment"),
		primary_action(values) {
			const rows = modes
				.map((mode, i) => ({
					mode_of_payment: mode.mode_of_payment,
					amount: flt(values[`amount_${i}`]),
				}))
				.filter((row) => row.amount > 0);

			if (!rows.length) {
				frappe.msgprint(__("Enter an amount against at least one mode of payment."));
				return;
			}

			const total = rows.reduce((sum, row) => sum + row.amount, 0);
			if (total - outstanding > 0.01) {
				frappe.msgprint(
					__("Tendered {0} is more than the {1} due.", [money(total), money(outstanding)])
				);
				return;
			}

			dialog.get_primary_btn().prop("disabled", true);
			frappe.call({
				method: "yht_custom.api.payment_assist.collect_payment",
				args: { sales_invoice: frm.doc.name, payments: rows },
				freeze: true,
				freeze_message: __("Recording payment…"),
				callback(r) {
					dialog.hide();
					const created = r.message || [];
					frappe.show_alert(
						{
							message:
								created.length === 1
									? __("Payment {0} recorded.", [created[0]])
									: __("{0} payments recorded.", [created.length]),
							indicator: "green",
						},
						7
					);
					frm.reload_doc();
				},
				error() {
					dialog.get_primary_btn().prop("disabled", false);
				},
			});
		},
		secondary_action_label: __("Pay Later"),
		secondary_action() {
			dialog.hide();
		},
	});

	// Running total, so a split is checkable before it is committed.
	const refresh_summary = () => {
		const total = modes.reduce(
			(sum, mode, i) => sum + flt(dialog.get_value(`amount_${i}`)),
			0
		);
		const balance = flt(outstanding - total, 2);
		const colour = balance < -0.01 ? "red" : balance > 0.01 ? "orange" : "green";
		dialog.fields_dict.summary.$wrapper.html(`
			<div style="display:flex;justify-content:space-between;font-weight:600">
				<span>${__("Tendered")}</span><span>${money(total)}</span>
			</div>
			<div style="display:flex;justify-content:space-between;color:var(--text-muted)">
				<span>${balance < 0 ? __("Over by") : __("Still due")}</span>
				<span style="color:var(--${colour}-500, inherit)">${money(Math.abs(balance))}</span>
			</div>
		`);
	};

	modes.forEach((mode, i) => {
		dialog.fields_dict[`amount_${i}`].df.onchange = refresh_summary;
	});

	dialog.show();

	// A single-mode branch has exactly one sensible answer — prefill it, and the
	// operator either confirms or edits. With several modes, prefilling the first
	// would be a guess.
	if (modes.length === 1) {
		dialog.set_value("amount_0", outstanding);
	}
	refresh_summary();
};
